// @prisma/adapter-planetscale/dist/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name2 in all)
    __defProp(target, name2, { get: all[name2], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  PrismaPlanetScale: () => PrismaPlanetScaleAdapterFactory
});
module.exports = __toCommonJS(index_exports);

// src/planetscale.ts
var planetScale = __toESM(require("@planetscale/database"));
var import_driver_adapter_utils2 = require("@prisma/driver-adapter-utils");
var import_async_mutex = require("async-mutex");

// package.json
var name = "@prisma/adapter-planetscale";

// src/conversion.ts
var import_database = require("@planetscale/database");
var import_driver_adapter_utils = require("@prisma/driver-adapter-utils");

// src/text.ts
var decoder = new TextDecoder("utf-8");
function decodeUtf8(text) {
  return text ? decoder.decode(uint8Array(text)) : "";
}
function uint8Array(text) {
  return Uint8Array.from(bytes(text));
}
function bytes(text) {
  return text.split("").map((c) => c.charCodeAt(0));
}

// src/conversion.ts
var BINARY_COLLATION_INDEX = 63;
function fieldToColumnType(field) {
  const type = field.type;
  switch (type) {
    case "INT8":
    case "UINT8":
    case "INT16":
    case "UINT16":
    case "INT24":
    case "UINT24":
    case "INT32":
    case "YEAR":
      return import_driver_adapter_utils.ColumnTypeEnum.Int32;
    case "UINT32":
    case "INT64":
    case "UINT64":
      return import_driver_adapter_utils.ColumnTypeEnum.Int64;
    case "FLOAT32":
      return import_driver_adapter_utils.ColumnTypeEnum.Float;
    case "FLOAT64":
      return import_driver_adapter_utils.ColumnTypeEnum.Double;
    case "TIMESTAMP":
    case "DATETIME":
      return import_driver_adapter_utils.ColumnTypeEnum.DateTime;
    case "DATE":
      return import_driver_adapter_utils.ColumnTypeEnum.Date;
    case "TIME":
      return import_driver_adapter_utils.ColumnTypeEnum.Time;
    case "DECIMAL":
      return import_driver_adapter_utils.ColumnTypeEnum.Numeric;
    case "CHAR":
    case "TEXT":
    case "VARCHAR":
      return import_driver_adapter_utils.ColumnTypeEnum.Text;
    case "ENUM":
      return import_driver_adapter_utils.ColumnTypeEnum.Enum;
    case "JSON":
      return import_driver_adapter_utils.ColumnTypeEnum.Json;
    case "BLOB":
    case "BINARY":
    case "VARBINARY":
      if (field.charset && field.charset !== BINARY_COLLATION_INDEX) {
        return import_driver_adapter_utils.ColumnTypeEnum.Text;
      }
      return import_driver_adapter_utils.ColumnTypeEnum.Bytes;
    case "BIT":
    case "BITNUM":
    case "HEXNUM":
    case "HEXVAL":
    case "GEOMETRY":
      return import_driver_adapter_utils.ColumnTypeEnum.Bytes;
    case "NULL":
      return import_driver_adapter_utils.ColumnTypeEnum.Int32;
    default:
      throw new Error(`Unsupported column type: ${type}`);
  }
}
var cast = (field, value) => {
  if (field.type === "JSON") {
    return typeof value === "string" && value.length ? decodeUtf8(value) : value;
  }
  return (0, import_database.cast)(field, value);
};
function mapArg(arg, argType) {
  if (arg === null) {
    return null;
  }
  if (typeof arg === "string" && argType.scalarType === "bigint") {
    return BigInt(arg);
  }
  if (argType.scalarType === "datetime" && typeof arg === "string") {
    arg = new Date(arg);
  }
  if (arg instanceof Date) {
    switch (argType.dbType) {
      case "TIME":
      case "TIME2":
        return formatTime(arg);
      case "DATE":
      case "NEWDATE":
        return formatDate(arg);
      default:
        return formatDateTime(arg);
    }
  }
  if (typeof arg === "string" && argType.scalarType === "bytes") {
    return Buffer.from(arg, "base64");
  }
  return arg;
}
function formatDateTime(date) {
  const pad = (n, z = 2) => String(n).padStart(z, "0");
  const ms = date.getUTCMilliseconds();
  return pad(date.getUTCFullYear(), 4) + "-" + pad(date.getUTCMonth() + 1) + "-" + pad(date.getUTCDate()) + " " + pad(date.getUTCHours()) + ":" + pad(date.getUTCMinutes()) + ":" + pad(date.getUTCSeconds()) + (ms ? "." + String(ms).padStart(3, "0") : "");
}
function formatDate(date) {
  const pad = (n, z = 2) => String(n).padStart(z, "0");
  return pad(date.getUTCFullYear(), 4) + "-" + pad(date.getUTCMonth() + 1) + "-" + pad(date.getUTCDate());
}
function formatTime(date) {
  const pad = (n, z = 2) => String(n).padStart(z, "0");
  const ms = date.getUTCMilliseconds();
  return pad(date.getUTCHours()) + ":" + pad(date.getUTCMinutes()) + ":" + pad(date.getUTCSeconds()) + (ms ? "." + String(ms).padStart(3, "0") : "");
}

// src/deferred.ts
function createDeferred() {
  const deferred = {};
  return [
    deferred,
    new Promise((resolve, reject) => {
      deferred.resolve = resolve;
      deferred.reject = reject;
    })
  ];
}

// src/errors.ts
function convertDriverError(error) {
  return {
    originalCode: `${error.code}`,
    originalMessage: error.message,
    ...mapDriverError(error)
  };
}
function mapDriverError(error) {
  switch (error.code) {
    case 1062: {
      const index = error.message.split(" ").pop()?.split("'").at(1)?.split(".").pop();
      return {
        kind: "UniqueConstraintViolation",
        constraint: index !== void 0 ? { index } : void 0
      };
    }
    case 1451:
    case 1452: {
      const field = error.message.split(" ").at(17)?.split("`").at(1);
      return {
        kind: "ForeignKeyConstraintViolation",
        constraint: field !== void 0 ? { fields: [field] } : void 0
      };
    }
    case 1263: {
      const index = error.message.split(" ").pop()?.split("'").at(1);
      return {
        kind: "NullConstraintViolation",
        constraint: index !== void 0 ? { index } : void 0
      };
    }
    case 1264:
      return {
        kind: "ValueOutOfRange",
        cause: error.message
      };
    case 1364:
    case 1048: {
      const field = error.message.split(" ").at(1)?.split("'").at(1);
      return {
        kind: "NullConstraintViolation",
        constraint: field !== void 0 ? { fields: [field] } : void 0
      };
    }
    case 1049: {
      const db = error.message.split(" ").pop()?.split("'").at(1);
      return {
        kind: "DatabaseDoesNotExist",
        db
      };
    }
    case 1007: {
      const db = error.message.split(" ").at(3)?.split("'").at(1);
      return {
        kind: "DatabaseAlreadyExists",
        db
      };
    }
    case 1044: {
      const db = error.message.split(" ").pop()?.split("'").at(1);
      return {
        kind: "DatabaseAccessDenied",
        db
      };
    }
    case 1045: {
      const user = error.message.split(" ").at(4)?.split("@").at(0)?.split("'").at(1);
      return {
        kind: "AuthenticationFailed",
        user
      };
    }
    case 1146: {
      const table = error.message.split(" ").at(1)?.split("'").at(1)?.split(".").pop();
      return {
        kind: "TableDoesNotExist",
        table
      };
    }
    case 1054: {
      const column = error.message.split(" ").at(2)?.split("'").at(1);
      return {
        kind: "ColumnNotFound",
        column
      };
    }
    case 1406: {
      const column = error.message.split(" ").flatMap((part) => part.split("'")).at(6);
      return {
        kind: "LengthMismatch",
        column
      };
    }
    case 1191:
      return {
        kind: "MissingFullTextSearchIndex"
      };
    case 1213:
      return {
        kind: "TransactionWriteConflict"
      };
    case 1040:
    case 1203:
      return {
        kind: "TooManyConnections",
        cause: error.message
      };
    default:
      return {
        kind: "mysql",
        code: error.code,
        message: error.message,
        state: error.state
      };
  }
}

// src/planetscale.ts
var debug = (0, import_driver_adapter_utils2.Debug)("prisma:driver-adapter:planetscale");
var RollbackError = class _RollbackError extends Error {
  constructor() {
    super("ROLLBACK");
    this.name = "RollbackError";
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, _RollbackError);
    }
  }
};
var PlanetScaleQueryable = class {
  constructor(client) {
    this.client = client;
  }
  provider = "mysql";
  adapterName = name;
  /**
   * Execute a query given as SQL, interpolating the given parameters.
   */
  async queryRaw(query) {
    const tag = "[js::query_raw]";
    debug(`${tag} %O`, query);
    const { fields, insertId: lastInsertId, rows } = await this.performIO(query);
    const columns = fields.map((field) => field.name);
    return {
      columnNames: columns,
      columnTypes: fields.map((field) => fieldToColumnType(field)),
      rows,
      lastInsertId
    };
  }
  /**
   * Execute a query given as SQL, interpolating the given parameters and
   * returning the number of affected rows.
   * Note: Queryable expects a u64, but napi.rs only supports u32.
   */
  async executeRaw(query) {
    const tag = "[js::execute_raw]";
    debug(`${tag} %O`, query);
    return (await this.performIO(query)).rowsAffected;
  }
  /**
   * Run a query against the database, returning the result set.
   * Should the query fail due to a connection error, the connection is
   * marked as unhealthy.
   */
  async performIO(query) {
    const { sql, args } = query;
    try {
      const result = await this.client.execute(
        sql,
        args.map((arg, i) => mapArg(arg, query.argTypes[i])),
        {
          as: "array",
          cast
        }
      );
      return result;
    } catch (e) {
      const error = e;
      onError(error);
    }
  }
};
function onError(error) {
  if (error.name === "DatabaseError") {
    const parsed = parseErrorMessage(error.message);
    if (parsed) {
      throw new import_driver_adapter_utils2.DriverAdapterError(convertDriverError(parsed));
    }
  }
  debug("Error in performIO: %O", error);
  throw error;
}
function parseErrorMessage(error) {
  const regex = /^(.*) \(errno (\d+)\) \(sqlstate ([A-Z0-9]+)\)/;
  let match = null;
  while (true) {
    const result = error.match(regex);
    if (result === null) {
      break;
    }
    match = result;
    error = match[1];
  }
  if (match !== null) {
    const [, message, codeAsString, sqlstate] = match;
    const code = Number.parseInt(codeAsString, 10);
    return {
      message,
      code,
      state: sqlstate
    };
  } else {
    return void 0;
  }
}
var PlanetScaleTransaction = class extends PlanetScaleQueryable {
  constructor(tx, options, txDeferred, txResultPromise) {
    super(tx);
    this.options = options;
    this.txDeferred = txDeferred;
    this.txResultPromise = txResultPromise;
  }
  // The PlanetScale connection objects are not meant to be used concurrently,
  // so we override the `performIO` method to synchronize access to it with a mutex.
  // See: https://github.com/mattrobenolt/ps-http-sim/issues/7
  #mutex = new import_async_mutex.Mutex();
  async performIO(query) {
    const release = await this.#mutex.acquire();
    try {
      return await super.performIO(query);
    } catch (e) {
      onError(e);
    } finally {
      release();
    }
  }
  async commit() {
    debug(`[js::commit]`);
    this.txDeferred.resolve();
    return await this.txResultPromise;
  }
  async rollback() {
    debug(`[js::rollback]`);
    this.txDeferred.reject(new RollbackError());
    return await this.txResultPromise;
  }
  async createSavepoint(name2) {
    await this.executeRaw({ sql: `SAVEPOINT ${name2}`, args: [], argTypes: [] });
  }
  async rollbackToSavepoint(name2) {
    await this.executeRaw({ sql: `ROLLBACK TO ${name2}`, args: [], argTypes: [] });
  }
  async releaseSavepoint(name2) {
    await this.executeRaw({ sql: `RELEASE SAVEPOINT ${name2}`, args: [], argTypes: [] });
  }
};
var PrismaPlanetScaleAdapter = class extends PlanetScaleQueryable {
  constructor(client) {
    super(client);
  }
  executeScript(_script) {
    throw new Error("Not implemented yet");
  }
  getConnectionInfo() {
    const url = this.client.connection()["url"];
    const dbName = new URL(url).pathname.slice(1);
    return {
      schemaName: dbName || void 0,
      // If `dbName` is an empty string, do not set a schema name
      supportsRelationJoins: true
    };
  }
  async startTransaction(isolationLevel) {
    const options = {
      usePhantomQuery: true
    };
    const tag = "[js::startTransaction]";
    debug("%s options: %O", tag, options);
    const conn = this.client.connection();
    if (isolationLevel) {
      await conn.execute(`SET TRANSACTION ISOLATION LEVEL ${isolationLevel}`).catch((error) => onError(error));
    }
    return this.startTransactionInner(conn, options);
  }
  async startTransactionInner(conn, options) {
    return new Promise((resolve, _reject) => {
      const txResultPromise = conn.transaction(async (tx) => {
        const [txDeferred, deferredPromise] = createDeferred();
        const txWrapper = new PlanetScaleTransaction(tx, options, txDeferred, txResultPromise);
        resolve(txWrapper);
        return deferredPromise;
      }).catch((error) => {
        if (!(error instanceof RollbackError)) {
          throw error;
        }
        return void 0;
      });
    });
  }
  async dispose() {
  }
  underlyingDriver() {
    return this.client;
  }
};
var PrismaPlanetScaleAdapterFactory = class {
  provider = "mysql";
  adapterName = name;
  #config;
  #client;
  constructor(arg) {
    if (arg instanceof planetScale.Client) {
      this.#config = arg.config;
      this.#client = arg;
    } else {
      this.#config = arg;
    }
  }
  async connect() {
    return new PrismaPlanetScaleAdapter(this.#client ?? new planetScale.Client(this.#config));
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  PrismaPlanetScale
});
